include/zephyr/pm/policy.h

```diff
@@ -275,63 +267,140 @@ static inline int32_t pm_policy_next_event_ticks(void)
+ * notification.
+ *
  * @param req Latency request.
  * @param value_us Maximum allowed latency in microseconds.
+ *
+ * @retval 0 if request is applied.
+ * @retval 1 if request required immediate action that is not completed. Configured completion
+ * notification will inform about completion.
+ * @retval -EIO immediate action returned error.
+ */
+int pm_policy_latency_request_add(struct pm_policy_latency_request *req, uint32_t value_us);
+
+/**
+ * @brief Synchronously add a new latency requirement.
+ *
+ * If immediate control manager is not used then it is an equivalent of
+ * @ref pm_policy_latency_request_add. If immediate control manager is used then context
+ * blocks until request is completed. It can only be called from the thread context.
+ * Asynchronous notification field in the @p req is configured and used internally.
+ *
+ * @retval 0 if request required immediate action that is not completed. Configured completion
+ * notification will inform about completion.
+ * @retval -EIO immediate action returned error.
  */
-void pm_policy_latency_request_add(struct pm_policy_latency_request *req,
-				   uint32_t value_us);
+int pm_policy_latency_request_add_sync(struct pm_policy_latency_request *req, uint32_t value_us);
 
 /**
  * @brief Update a latency requirement.
  *
  * @param req Latency request.
  * @param value_us New maximum allowed latency in microseconds.
+ *
+ * @retval 0 if request is applied.
+ * @retval 1 if request required immediate action that is not completed. Configured completion
+ * notification will inform about completion.
+ * @retval -EIO immediate action returned error.
+ */
+int pm_policy_latency_request_update(struct pm_policy_latency_request *req, uint32_t value_us);
+
+/**
+ * @brief Synchronously update a latency requirement.
+ *
+ * If immediate control manager is not used then it is an equivalent of
+ * @ref pm_policy_latency_request_update. If immediate control manager is used then context
+ * blocks until request is completed. It can only be called from the thread context.
+ * Asynchronous notification field in the @p req is configured and used internally.
+ *
+ * @retval 0 if request required immediate action that is not completed. Configured completion
+ * notification will inform about completion.
+ * @retval -EIO immediate action returned error.
  */
-void pm_policy_latency_request_update(struct pm_policy_latency_request *req,
-				      uint32_t value_us);
+int pm_policy_latency_request_update_sync(struct pm_policy_latency_request *req, uint32_t value_us);
 
 /**
  * @brief Remove a latency requirement.
  *
  * @param req Latency request.
+ *
+ * @retval 0 if request removed successfully.
+ * @retval -EALREADY if request is not active and cannot be removed.
+ * @retval -EIO immediate action returned error.
+ */
+int pm_policy_latency_request_remove(struct pm_policy_latency_request *req);
+
+/** @brief Immediate action manager for single threshold.
```

Comment:
The "immediate" word used around max latency changes subscription is confusing to me. Is there any "delayed max latency changes subscription" to differentiate the "immediate" one from?

Maybe it would be enough to name this structure `pm_policy_latency_change_binary_subscription`?

Reply:
I don't fully agree. When latency is used to select a CPU sleep state, the immediate action is locking a set of CPU sleep states that are characterized by greater latency. I agree, that the visible effect of this action is only when the CPU enters the idle thread, but we can treat locking as an immediate action.

The same way with NVM controllers: the immediate action is locking their sleep states that are characterized by greater latency. But this action has visible effect after the NVM inactivity timeout fires.

So, you can treat both of these actions as "immediate" - locking states are immediate in both, or as "deferred" - the actual sleep states are entered after inactivity is detected.

The main difference between these two is that CPU-idle handling is CPU-specific, while NVM-idle handling is SoC-specific or even PCB-specific if an NVM controller is external.